sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/utils/CalciteUtils.java

```diff
@@ -154,6 +163,7 @@ public static boolean isStringType(FieldType fieldType) {
           .put(TIME_WITH_LOCAL_TZ, SqlTypeName.TIME_WITH_LOCAL_TIME_ZONE)
           .put(TIMESTAMP, SqlTypeName.TIMESTAMP)
           .put(TIMESTAMP_WITH_LOCAL_TZ, SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE)
+          .put(DATETIME, SqlTypeName.TIMESTAMP)
```

Comment:
The challenging here is how to choose the SqlTypeName in Calcite for DateTime. As there is not such SqlTypeName exist for DateTime, a workaround is use `TIMESTAMP_WITH_LOCAL_TIME_ZONE` for DateTime.

Reply:
This seems to be the crux of the issue. This came from BigQuery? If I'm reading the docs correctly BigQuery `DATETIME` is what the SQL standard calls `TIMESTAMP [WITHOUT TIME ZONE]` (and Calcite calls `TIMESTAMP`). BigQuery `TIMESTAMP` is what the SQL standard calls `TIMESTAMP WITH TIME ZONE` (and Calcite calls [`TIMESTAMP_WITH_TIMEZONE`](https://github.com/apache/calcite/blob/9721283bd0ce46a337f51a3691585cca8003e399/core/src/main/java/org/apache/calcite/sql/type/SqlTypeFamily.java#L104) and maps to `TIMESTAMP`).

I think the mapping to `TIMESTAMP` is correct here but is this actually just `TIMESTAMP`? What is the benefit we get by adding an extra type?

Reply:
Because if we have both `DATETIME`(`TIMESTAMP[WITHOUT TIME ZONE]`) and `TIMESTAMP`(`TIMESTAMP WITH TIME ZONE`) mapped to `TIMESTAMP`, we won't be able to differentiate these two types in Calcite logical plans(For example, schema is created from RowType in BeamCalcRel: https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java#L122)

Reply:
If you map to `TIMESTAMP_WITH_LOCAL_TIME_ZONE` as a workaround, you end up with the same problem just with different data types. It seems that differentiating between the two types is not possible without a change to calcite.

Reply:
(Or a change to how we interact with calcite.)

Reply:
For sure ideally the change should happen to Calcite (add the corresponding type for example).

As a workaround, what I was really thinking is  use `TIMESTAMP_WITH_LOCAL_TIME_ZONE` for `DateTime` and we actually disable `TIMESTAMP_WITH_LOCAL_TIME_ZONE` as a type for BeamSQL right now. If you think this is not acceptable, then I guess we have to make a change to Calcite.

Reply:
What are the differences between DATETIME and TIMESTAMP WITHOUT TIME ZONE, so that they need to be distinguished?

Reply:
I cannot find a precious definition of `TIMESTMAP WITHOUT TIME ZONE` from Calcite so I don't really know the difference.

However Calcite maps both `TIMESTMAP WITHOUT TIME ZONE` and ``TIMESTMAP WITH TIME ZONE` to `TIMESTAMP`

Reply:
The problem is Beam ZetaSQL will only see Calcite Schema. If we treat `DATETIME` as Calcite's `TIMESTAMP` (thus mixed `DATETIME` with `TIMESTAMP WITH TIME ZONE`), it means Beam ZetaSQL loses the information of `DATETIME` type in plans.

Reply:
I've done some reading and I agree that BQ DATETIME is the same as TIMESTAMP WITHOUT TIME ZONE. The conversion is lossless except for precision on the subsecond part. BQ TIMESTAMP is the same as TIMESTAMP WITH TIME ZONE.

The current mapping here is incorrect.

Reply:
Yes. This mapping is not correct.

The discussion here is what mapping we should use for DATETIME. If you check SqlTypename, you will find there is no TIMESTAMP without TIME ZONE (or equivalent type) exist.

Reply:
I believe that `TIMESTAMP` is already without time zone. See https://issues.apache.org/jira/browse/CALCITE-2394 for some discussion I have had about some of this.

Reply:
It is interesting then. I haven't read the relevant JIRA but at least sounds like at least one of our mapping is wrong, either DATETIME or TIMESTAMP WITH TIME ZONE.